scipy/spatial/transform/rotation.py

```diff
@@ -8,7 +8,107 @@
 AXIS_TO_IND = {'x': 0, 'y': 1, 'z': 2}
 
 
-def _make_elementary_quat(axis, angles):
+def elementary_basis_vector(axis):
+    b = np.zeros(3)
+    b[AXIS_TO_IND[axis]] = 1
+    return b
+
+
+def compute_euler_from_dcm(dcm, seq, extrinsic=False):
+    # The algorithm assumes intrinsic frame transformations. For representation
+    # the paper uses transformation matrices, which are transpose of the
+    # direction cosine matrices used by our Rotation class.
+    # Adapt the algorithm for our case by
+    # 1. Transposing our dcm representation
+    # 2. Reversing both axis sequence and angles for extrinsic rotations
+
+    if extrinsic:
+        seq = seq[::-1]
+
+    if dcm.ndim == 2:
+        dcm = dcm[None, :, :]
+    num_rotations = dcm.shape[0]
+
+    # Step 0
+    # Algorithms dcms are transpose of ours
+    dcm = np.transpose(dcm, (0, 2, 1))
+    # Algorithm assumes axes as column vectors, here we use 1D vectors
+    n1 = elementary_basis_vector(seq[0])
+    n2 = elementary_basis_vector(seq[1])
+    n3 = elementary_basis_vector(seq[2])
+
+    # Step 2
+    sl = np.dot(np.cross(n1, n2), n3)
+    cl = np.dot(n1, n3)
+
+    # angle phi is lambda from the paper referenced in [2] from docstring of
+    # `as_euler` function
+    phi = np.arctan2(sl, cl)
+    c = np.empty((3, 3))
+    c[0] = n2
+    c[1] = np.cross(n1, n2)
+    c[2] = n1
+
+    # Step 3
+    rt = np.array([
+        [1, 0, 0],
+        [0, cl, -sl],
+        [0, sl, cl]
+    ])
+    rtc = rt.dot(c)
+    ct = c.T
+    res = np.einsum('...ij,...jk->...ik', rtc, dcm)
+    o = np.einsum('...ij,...jk->...ik', res, ct)
+
+    # Step 4
+    offset = np.arccos(o[:, 2, 2])
+    angle2 = phi + offset
+
+    # Steps 5, 6
+    eps = np.finfo(float).resolution  # ~1e-15
+    safe1 = (np.abs(offset) >= eps)
+    safe2 = (np.abs(offset - np.pi) >= eps)
+
+    angle1 = np.empty(num_rotations)
+    angle3 = np.empty(num_rotations)
+
+    # 5b
+    safe_mask = np.logical_and(safe1, safe2)
+    angle1[safe_mask] = np.arctan2(o[safe_mask, 2, 0], -o[safe_mask, 2, 1])
+    angle3[safe_mask] = np.arctan2(o[safe_mask, 0, 2], o[safe_mask, 1, 2])
+
+    # 6a
+    angle3[~safe_mask] = 0
+    # 6b
+    angle1[~safe1] = np.arctan2(o[~safe1, 0, 1] - o[~safe1, 1, 0],
+                                o[~safe1, 0, 0] + o[~safe1, 1, 1])
+    # 6c
+    angle1[~safe2] = np.arctan2(o[~safe2, 0, 1] + o[~safe2, 1, 0],
+                                o[~safe2, 0, 0] - o[~safe2, 1, 1])
+
+    # Step 7
+    # python modulo operator works correctly for negative numbers
+    if seq[0] == seq[2]:
+        # lambda = 0, so we can only ensure angle2 -> [0, pi]
+        adjust_mask = np.logical_or(angle2 < 0, angle2 > np.pi)
+    else:
+        adjust_mask = np.logical_or(angle2 < -np.pi / 2, angle2 > np.pi / 2)
```

Comment:
Fo this case, what are possible values for `phi`? +pi/2 or -pi/2?

Reply:
Yes, I'll put that as a comment.